### PR TITLE
SelectMultiple - Change select all and clear all buttons to show when searching

### PR DESCRIPTION
**Pull request by jcfilben**
#### What does this PR do?
Adjusts the behavior of SelectMultiple to show 'Select All' and 'Clear All' buttons when searching. 

Note: In the original issue it was mentioned that the buttons should be labeled 'Select All Matches' and 'Clear All Matches' when the options are being filtered by a search term. When I tried this the button labels felt too long and they were taking up too much space within the SelectMultiple drop.

This PR also fixes an issue where selected options were bubbling up when searching. Options should only bubble on after the drop is closed an re-opened.

#### How should this be manually tested?
Test with the existing SelectMultiple stories. Open the drop and type in the search. The 'Select All'/'Clear All' button should still be visible.

#### Do Jest tests follow these best practices?

- [ ] `screen` is used for querying.
- [ ] The correct query is used. (Refer to [this list of queries](https://testing-library.com/docs/queries/about/#priority))
- [ ] `userEvent` is used in place of `fireEvent`.
- [ ] `asFragment()` is used for snapshot testing.

#### What are the relevant issues?
Closes https://github.com/grommet/grommet/issues/6444
Closes https://github.com/grommet/grommet/issues/6447

#### Do the grommet docs need to be updated?
No
#### Should this PR be mentioned in the release notes?
Yes
#### Is this change backwards compatible or is it a breaking change?
Backwards compatible

### CODE REVIEW

**Review by halocline — COMMENTED** [review]
There is a scenario where I :
- search for "p"
- Select All selects "Apple", "Grape", and "Raspberry"
- clear search and search for "z"

Do we show a button here? I don't think so, but just asking to validate our current thinking.

-----------

There is a scenario where I :
- search for "p"
- Select All selects "Apple", "Grape", and "Raspberry"
- clear search and search for "w"
- Select All adds "Strawberry" and "Kiwi" to the existing selection, resulting in 5 options selected
- clear search and search for "b"

We show "Clear all" instead of "Select all" here. Is that what we want? I think so, but again just surveying edge cases.

------------

Lastly, I think this PR would be a good candidate to improve our test scenario coverage. Can we add a Jest test which executes the second scenario above with assertions throughout?

**Comment by jcfilben** [comment]
> There is a scenario where I :
> 
> search for "p"
> Select All selects "Apple", "Grape", and "Raspberry"
> clear search and search for "z"
> Do we show a button here? I don't think so, but just asking to validate our current thinking.

We don't show a button here since no matches were found after searching 'z'.

> There is a scenario where I :
> 
> search for "p"
> Select All selects "Apple", "Grape", and "Raspberry"
> clear search and search for "w"
> Select All adds "Strawberry" and "Kiwi" to the existing selection, resulting in 5 options selected
> clear search and search for "b"
> We show "Clear all" instead of "Select all" here. Is that what we want? I think so, but again just surveying edge cases.

Yes, we show 'Clear All' whenever any options are in the search results and selected.

> Lastly, I think this PR would be a good candidate to improve our test scenario coverage. Can we add a Jest test which executes the second scenario above with assertions throughout?

Good idea I realized I didn't add a test in my latest commit but I will go back and add a test for this

File: src/js/components/SelectMultiple/SelectMultiple.js [hunk 1]
```diff
@@ -144,7 +144,7 @@ const SelectMultiple = forwardRef(
     }, [optionsProp, search]);
 
     useEffect(() => {
-      if (!search && sortSelectedOnClose) setOrderedOptions(optionsProp);
+      if (sortSelectedOnClose) setOrderedOptions(optionsProp);
     }, [optionsProp, search, sortSelectedOnClose]);
```
[the remark this answers is not in the document]
Comment (jcfilben) [reply src/js/components/SelectMultiple/SelectMultiple.js]: Fixed in latest commit

File: src/js/components/SelectMultiple/SelectMultiple.js [hunk 2]
```diff
@@ -168,7 +168,7 @@ const SelectMultiple = forwardRef(
     }, [onOpen, open]);
 
     useEffect(() => {
```
[the remark this answers is not in the document]
Comment (jcfilben) [reply src/js/components/SelectMultiple/SelectMultiple.js:172]: Added a comment in latest commit

File: src/js/components/SelectMultiple/SelectMultiple.js [hunk 3]
```diff
@@ -168,7 +168,7 @@ const SelectMultiple = forwardRef(
...
-      if (sortSelectedOnClose && ((open && search) || !open)) {
+      if (sortSelectedOnClose && value && !open) {
```
[the remark this answers is not in the document]
Comment (jcfilben) [reply src/js/components/SelectMultiple/SelectMultiple.js:173]: Yup, done in latest commit

File: src/js/components/SelectMultiple/SelectionSummary.js [hunk 4]
```diff
@@ -53,70 +53,86 @@ const SelectionSummary = ({
...
+      flex={showSelectedInline}
+    >
+      <Box pad={{ vertical: 'xsmall' }} alignSelf="center">
+        <Text size="small">
+          {value?.length === 0 ||
+          onMore ||
+          !value ||
+          (search !== '' && search !== undefined)
+            ? `${value?.length || 0} selected`
+            : `${value?.length || 0} selected of ${options.length}`}
```
[the remark this answers is not in the document]
Comment (jcfilben) [reply src/js/components/SelectMultiple/SelectionSummary.js]: Done

File: src/js/components/SelectMultiple/SelectionSummary.js [hunk 5]
```diff
@@ -53,70 +53,86 @@ const SelectionSummary = ({
...
+      pad={
+        showSelectedInline ? { left: 'xsmall', vertical: 'xsmall' } : 'xsmall'
+      }
+      direction="row"
+      justify="between"
+      gap="small"
+      fill="horizontal"
+      flex={showSelectedInline}
+    >
+      <Box pad={{ vertical: 'xsmall' }} alignSelf="center">
```
[the remark this answers is not in the document]
Comment (jcfilben) [reply src/js/components/SelectMultiple/SelectionSummary.js]: Good note, I removed the extra Box and added `height={{ min: 'xxsmall' }}` to the parent

File: src/js/components/SelectMultiple/SelectionSummary.js [hunk 6]
```diff
@@ -53,70 +53,86 @@ const SelectionSummary = ({
...
+        0 &&
+        (!onMore || (onMore && value?.length !== 0)) && (
+          <Button
+            a11yTitle={
+              showSelectAll
+                ? `Select all ${options.length} options`
+                : `${value?.length} options selected. Clear all?`
+            }
+            label={showSelectAll ? 'Select All' : 'Clear All'}
+            onClick={(event) => {
```
[the remark this answers is not in the document]
Comment (jcfilben) [reply src/js/components/SelectMultiple/SelectionSummary.js]: Yup, done

File: src/js/components/SelectMultiple/SelectionSummary.js [hunk 7]
```diff
@@ -53,70 +53,86 @@ const SelectionSummary = ({
...
+        <Text size="small">
+          {value?.length === 0 ||
+          onMore ||
+          !value ||
+          (search !== '' && search !== undefined)
+            ? `${value?.length || 0} selected`
+            : `${value?.length || 0} selected of ${options.length}`}
+        </Text>
+      </Box>
+      {(options.length &&
```
[the remark this answers is not in the document]
Comment (jcfilben) [reply src/js/components/SelectMultiple/SelectionSummary.js:123]: I think the ideal behavior is not showing a button when we have no options

File: src/js/components/SelectMultiple/SelectionSummary.js [hunk 8]
```diff
@@ -53,70 +53,86 @@ const SelectionSummary = ({
...
+          !(!value || (value?.length === 0 && selectedValuesDisabled())))) >
+        0 &&
+        (!onMore || (onMore && value?.length !== 0)) && (
+          <Button
+            a11yTitle={
+              showSelectAll
+                ? `Select all ${options.length} options`
+                : `${value?.length} options selected. Clear all?`
+            }
+            label={showSelectAll ? 'Select All' : 'Clear All'}
```
[the remark this answers is not in the document]
Comment (jcfilben) [reply src/js/components/SelectMultiple/SelectionSummary.js:134]: Matt and I had an offline discussion about this one and agreed it could be handled in a separate issue. Changing to "Select matches" and "Clear matches" is still affecting the width of the drop container. I think the naming is clearer but we may want to look into sizing of the button, how comfortable are we with affecting the drop container width, etc. So I think this work could be handled separately

**Comment by jcfilben** [comment]
Marking as draft, I need to go back and add a unit test

**Review by halocline — APPROVED** [review]

**Review by ericsoderberghp — APPROVED** [review]
